clang/lib/Lex/PPDirectives.cpp

```diff
@@ -3902,6 +3902,8 @@ void Preprocessor::HandleEmbedDirectiveImpl(
   }
 
   EmbedAnnotationData *Data = new (BP) EmbedAnnotationData;
+  Data->Filename = Filename;
+  Data->IsAngled = IsAngled;
   Data->BinaryData = BinaryContents;
```

Comment:
I struggle to see how the current implementation actually prevents the memory leak. `EmbedAnnotationData` is allocated first as if the `std::string` inside was empty, then the string is assigned with an array of chars which will likely extend string's size, giving no chance to deallocate proper size later. I wonder if keeping a `StringRef` in `EmbedAnnotationData` is possible instead.

Reply:
Is there such a thing as an OwningStringRef then? I am not sure how to guarantee the string exists, and then is deallocated along with the tokens. I thought std::string would be a ptr to the dynamically sized string data, but I haven't used the STL much

Reply:
None that I can think of... I don't think it would've helped though since I don't think the bump allocator calls the destructors.
I think we need to find a good place to store filename strings that come from embed directives. @AaronBallman any chance you have a good ideas about where can we store the filename string to manage the memory in a good way?

Reply:
Similar issue with the deserializer, not sure where to store the string.

Reply:
What if it's something like a flexible array member and the EmbedAnnotationData is allocated to the right size the first time?

Reply:
> where can we store the filename string to manage the memory in a good way?

Can’t we just copy the string data into the ASTContext’s allocator and store a StringRef to it, or am I missing something here?

Reply:
I'm not sure you can do that during the lexing stage

Reply:
I think the way this may need to work is to find out where we're getting the file name from and seeing how long-lived that string is. It's passed here as a `StringRef`, and it would nice if we can store it as a `StringRef` in the `EmbedAnnotationData` rather than as a `std::string`. My guess is that the name comes from the `FileManager` and is as long-lived as the `SourceManager` which is owned by `ASTContext`, but we'd need to verify that.